### PR TITLE
[binary filter] change boolean ros params

### PR DESCRIPTION
**Pull request by enricosutera**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   |  https://github.com/ros-planning/navigation2/issues/3511# |
| Primary OS tested on | Ubuntu 22.04, Rolling (debian) |
| Robotic platform tested on | None |

---

## Description of contribution in a few bullet points

* The binary filter is now able to change boolean ros parameters using a service. The parameters can be changed accordingly to filter binary value or opposite to it (this can be specified for each parameter).

## Description of documentation updates required from your changes

* Added new parameters for the binary filters, documentation TODO

## Future work that may be required in bullet points

* Not required, but an extension or another filter could  be developed to change also `double` parameters dynamically. 

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 1]
```diff
@@ -259,6 +295,61 @@ void BinaryFilter::changeState(const bool state)
...
+            bool_param.value.bool_value ? "true" : "false");
+        }
+      };
+
+    RCLCPP_DEBUG(
+      logger_, "BinaryFilter: Sending request to set parameter  %s to %s",
+      binary_parameters_info_.at(param_index).param_name.c_str(),
+      bool_param.value.bool_value ? "true" : "false");
+    auto future_result = change_parameters_clients_.at(param_index)->async_send_request(
+      request, response_received_callback);
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:351]: Remove callback. Spin until future complete. Then check the value and throw something more substantial than a log if it failed

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
I think also this entire thing could use a single https://docs.ros2.org/bouncy/api/rclcpp/classrclcpp_1_1_async_parameters_client.html, having the individual `/set_parameter` clients seems not great

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 2]
```diff
@@ -259,6 +295,61 @@ void BinaryFilter::changeState(const bool state)
...
+    }
+
+    // Create a rcl_interfaces::msg::SetParameters client for changing parameters
+    auto request = std::make_shared<rcl_interfaces::srv::SetParameters::Request>();
+
+    // Set parameters for BinaryFilter
+    rcl_interfaces::msg::Parameter bool_param;
+    bool_param.name = binary_parameters_info_.at(param_index).param_name;
+    bool_param.value.type = rcl_interfaces::msg::ParameterType::PARAMETER_BOOL;
+    if (binary_parameters_info_.at(param_index).default_state != default_state_) {
```
[the remark this answers is not in the document]
Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:324]: This way is some kind of tricky: the state of each parameter will flip in the same or opposite way depending on whether the parameter's initial default state will be equal to global filter state. So, this construction will work, but not it is obvious.

How about to change the logic a bit to express the intention more clearly:
```
    const bool param_default_state = binary_parameters_info_.at(param_index).default_state;
    if (state == default_state_) {
      // Filter is not flipped
      bool_param.value.bool_value = param_default_state;
    } else {
      // Filter is flipped
      bool_param.value.bool_value = !param_default_state;
    }
```

**Review by AlexeyMerzlyakov — COMMENTED** [review]
The main code has been checked. Please check the first Steve review and these items. After the code will be more-less stabilized, we could advance to next step with testcases and documentation verification.

**Comment by AlexeyMerzlyakov** [comment]
@enricosutera, this is kind ping, since there are no news more than one month. Are there any updates on this PR?

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 3]
```diff
@@ -73,6 +75,27 @@ void BinaryFilter::initializeFilter(
   node->get_parameter(name_ + "." + "binary_state_topic", binary_state_topic);
   declareParameter("flip_threshold", rclcpp::ParameterValue(50.0));
   node->get_parameter(name_ + "." + "flip_threshold", flip_threshold_);
+  declareParameter("binary_parameters", rclcpp::ParameterValue(std::vector<std::string>()));
```
[the remark this answers is not in the document]
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:76]: [Paramter description](https://github.com/enricosutera/navigation.ros.org/blob/36d105555a90f2cec641b47ed20e005f4e717682/configuration/packages/costmap-plugins/binary_filter.rst?plain=1#L92)

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 4]
```diff
@@ -73,6 +75,27 @@ void BinaryFilter::initializeFilter(
...
+  node->get_parameter(name_ + "." + "binary_parameters", binary_parameters_);
+  declareParameter("change_parameter_timeout", rclcpp::ParameterValue(10));
```
[the remark this answers is not in the document]
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:78]: [Parameter description](https://github.com/enricosutera/navigation.ros.org/blob/36d105555a90f2cec641b47ed20e005f4e717682/configuration/packages/costmap-plugins/binary_filter.rst?plain=1#L81C13-L81C13)

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 5]
```diff
@@ -259,6 +295,61 @@ void BinaryFilter::changeState(const bool state)
     std::make_unique<std_msgs::msg::Bool>();
   msg->data = state;
   binary_state_pub_->publish(std::move(msg));
+  changeParameters(state);
+}
+
+void BinaryFilter::changeParameters(const bool state)
+{
+  for (size_t param_index = 0; param_index < binary_parameters_info_.size(); ++param_index) {
+    if (!change_parameters_clients_.at(param_index)->wait_for_service(
```
[the remark this answers is not in the document]
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:302]: Also related to previous comment:

> Perform wait for service here to make sure up before continuing

Do you mean that if any service goes timeout at the beginning we should not add it or just give an error/warning? What if the server is temporary offline (e.g. from a plugin)? 
Also, a node could go offline at a certain point and since the service is checked only at initialization time, there would not be information for the user to understand that parameters are not updated.
I understand that adding the `wait_for_server` slows down the execution, but in my opinion it should be the user to keep attention to the selected parameters (e.g. too many "offline" servers).

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:322]: > How can you have a global default_state_ for potentially many unrelated parameters in unrelated nodes?

I think it's realistic to have heavy relations between parameters, given that the mask is shared. For instance one could set it to disable an instance of the obstacle layer and enable another one (and this may hold for other layers too). This way one would need to specify only the default value for half of the parameter, while the other half can be left default (as the global default state)

[the remark this answers is not in the document]
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp]: > This way is some kind of tricky: the state of each parameter will flip in the same or opposite way depending on whether the parameter's initial default state will be equal to global filter state. So, this construction will work, but not it is obvious.
> 
> How about to change the logic a bit to express the intention more clearly:
> ...

Thank you, much more readable!

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 1 above)
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:353]: I guess it depends on what you think about [this](https://github.com/ros-planning/navigation2/pull/3796#discussion_r1369208988)

**Comment by SteveMacenski** [comment]
@enricosutera good point, let me look at this again - but you have not yet responded to the bulk of the comments in my last review

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 6]
```diff
@@ -73,6 +73,26 @@ void BinaryFilter::initializeFilter(
...
   node->get_parameter(name_ + "." + "flip_threshold", flip_threshold_);
+  declareParameter("binary_parameters", rclcpp::ParameterValue(std::vector<std::string>()));
+  node->get_parameter(name_ + "." + "binary_parameters", binary_parameters_);
+  declareParameter("change_parameter_timeout", rclcpp::ParameterValue(10));
+  node->get_parameter(name_ + "." + "change_parameter_timeout", change_parameter_timeout_);
+
+  for (std::string param : binary_parameters_) {
+    BinaryParameter param_struct;
+
+    declareParameter(param + "." + "node_name", rclcpp::ParameterValue(""));
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:84]: If there's no default, don't set as empty, set as ParameterType::String so that it throws an exception if its not fully specified. You'd need to wrap parameter getting in try/catch to catch this case and log an error + report it

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 7]
```diff
@@ -259,6 +291,65 @@ void BinaryFilter::changeState(const bool state)
...
+        std::chrono::milliseconds(change_parameter_timeout_)))
+    {
+      RCLCPP_WARN(
+        logger_, "BinaryFilter:  service %s not available. Skipping ...",
+        change_parameters_clients_.at(param_index)->get_service_name());
+      continue;
+    } else {
+      RCLCPP_INFO(
+        logger_, "BinaryFilter:  service %s available.",
+        change_parameters_clients_.at(param_index)->get_service_name());
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:310]: Look ups aren't free. Store a reference to `change_parameters_clients_.at(param_index)` and use it once

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 8]
```diff
@@ -259,6 +291,65 @@ void BinaryFilter::changeState(const bool state)
...
+    }
+
+    // Create a rcl_interfaces::msg::SetParameters client for changing parameters
+    auto request = std::make_shared<rcl_interfaces::srv::SetParameters::Request>();
+
+    // Set parameters for BinaryFilter
+    rcl_interfaces::msg::Parameter bool_param;
+    bool_param.name = binary_parameters_info_.at(param_index).param_name;
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:318]: Ditto

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 9]
```diff
@@ -259,6 +291,65 @@ void BinaryFilter::changeState(const bool state)
...
+            logger_, "BinaryFilter: Successfully changed parameter to %s",
+            bool_param.value.bool_value ? "true" : "false");
+        }
+      };
+
+    RCLCPP_DEBUG(
+      logger_, "BinaryFilter: Sending request to set parameter  %s to %s",
+      binary_parameters_info_.at(param_index).param_name.c_str(),
+      bool_param.value.bool_value ? "true" : "false");
+    auto future_result = change_parameters_clients_.at(param_index)->async_send_request(
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:350]: If future_result goes out of scope at the end of the loop, does this actually log anything once its been destroyed?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by SteveMacenski** [comment]
But generally looks good!

**Comment by AlexeyMerzlyakov** [comment]
@enricosutera, thank you for the update and taking care on this. Started the review, but I'm running out-of-time today. Will complete it at the beginning of next week.

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:300]: Once the BF is being configured, we need to check that services/nodes are available on that stage. Later on we are assuming that everything is OK, and we are working with bringed-up nodes. If something goes wrong (e.g. node became offline), service request should fail by timeout. For that, see Steve comment from above and please use `spin_until_future_complete(future_result, change_parameter_timeout_)` API to not have infinite loops. Anyway, loosing the service with the node - is extraordinary situation, which could be handled by other than each time checking for service availability way (performance for usual serivce calls in this way will be faster, and we still could handle error situations).

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 10]
```diff
@@ -84,6 +104,18 @@ void BinaryFilter::initializeFilter(
...
     std::bind(&BinaryFilter::filterInfoCallback, this, std::placeholders::_1));
 
+  // Create clients for changing parameters
+  for (auto param : binary_parameters_info_) {
+    RCLCPP_DEBUG(
+      logger_,
+      "BinaryFilter: Creating client for changing parameter \"%s\" of node \"%s\"...",
+      param.param_name.c_str(), param.node_name.c_str());
+
+    auto change_parameters_client_ = node->create_client<rcl_interfaces::srv::SetParameters>(
```
Comment (AlexeyMerzlyakov) [inline nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:114]: Please remove `_` suffix at the end of variable name, since it is local

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_filters/binary_filter.hpp [hunk 11]
```diff
@@ -118,6 +125,26 @@ class BinaryFilter : public CostmapFilter
...
+    std::string node_name;    // Node to which the parameter belongs
+    std::string param_name;   // Name of parameter
+    bool default_state;       // Parameter default state
+  };
+
+  // List of params with info
+  std::vector<BinaryParameter> binary_parameters_info_;
+  // List of clients for changing parameters
+  std::vector<rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedPtr>
+  change_parameters_clients_;
```
Comment (AlexeyMerzlyakov) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_filters/binary_filter.hpp:144]: This will cause linting falures - please correct the indents:
```
  std::vector<rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedPtr>
    change_parameters_clients_;
```

**Review by AlexeyMerzlyakov — COMMENTED** [review]
There are some items remaining to fix from this and previous reviews - adding the review comments for the main code

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_filters/binary_filter.hpp (hunk 11 above)
Comment (enricosutera) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_filters/binary_filter.hpp:141]: To be honest I get the opposite result:
```
5: Code style divergence in file 'include/nav2_costmap_2d/costmap_filters/binary_filter.hpp':
5: 
5: --- include/nav2_costmap_2d/costmap_filters/binary_filter.hpp
5: +++ include/nav2_costmap_2d/costmap_filters/binary_filter.hpp.uncrustify
5: @@ -140,2 +140,2 @@
5: -  std::vector<rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedPtr> 
5: -    change_parameters_clients_;
5: +  std::vector<rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedPtr>
5: +  change_parameters_clients_;
5: 
5: 1 files with code style divergence
```

What am I missing?

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 12]
```diff
@@ -84,6 +107,19 @@ void BinaryFilter::initializeFilter(
...
+  // Create clients for changing parameters
+
+  for (auto param : binary_parameters_info_) {
+    RCLCPP_DEBUG(
+      logger_,
+      "BinaryFilter: Creating client for changing parameter \"%s\" of node \"%s\"...",
+      param.param_name.c_str(), param.node_name.c_str());
+
+    auto change_parameters_client_ = node->create_client<rcl_interfaces::srv::SetParameters>(
+      "/" + param.node_name + "/set_parameters");
```
[the remark this answers is not in the document]
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:117]: I can removed the absolute slash (`"/" +`) to make it automatically handle namespaces, but I guess then there would be an issue with `/local_costmap`/`/global_costmap` sub-namespace of the costmap, which - I think - is the same reason why you use `/scan` (instead of scan) in the obstacle layers (using a relative source, e.g. `scan` creates a subscription to `/global_costmap/scan`). 

I mean, I kept it absolute to have more control. This way a user can add the namespace in the params file, while if the namespace would be handled automatically a user would need to do a remap.
What do you suggest?
I think something like this (from[ costmap_2d_ros](https://github.com/ros-planning/navigation2/blob/d9bea52838d67b7ecfefb60788a3a4ff61f587ca/nav2_costmap_2d/src/costmap_2d_ros.cpp#L103) would solve it:
```
     "map_topic", rclcpp::ParameterValue(
      (parent_namespace_ == "/" ? "/" : parent_namespace_ + "/") + std::string("map")));

```

But I don't have such information (`parent_namespace_`) at costmap filter level

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 7 above)
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp]: https://github.com/ros-planning/navigation2/pull/3796/commits/66b006b6ad0fd80eada2ab529a6741917f9d1b51

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 8 above)
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp]: https://github.com/ros-planning/navigation2/pull/3796/commits/66b006b6ad0fd80eada2ab529a6741917f9d1b51

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp]: Hi there. I agree with using `spin_until_future_complete(future_result, change_parameter_timeout_)`. However to me it's still not clear what should the filter do if a node is offline. I mean, service request fail by timeout and then? Should I throw an error? While log seems not much substantial, as I was saying this seems to be a little bit much. What if a node is not active yet or has just crashed and will be up again in a few seconds?
Anyway, it is only a matter of choices. If you want this to work only with nodes that are always up, it's ok! 
By the way I could also add another parameter to somehow distinguish between "always-up" nodes and "discontinuous" ones (for instance plugin). Or adding a "global" timeout (not for the service request timeout but for the node in general, for instance 10 seconds timeout after the first service request timeout... this would implies some re-call logic, even when there's no binary state change). Or simply logging but also exposing error information for any third part diagnostic nodes... I really don't know, just shooting ideas

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp [hunk 13]
```diff
@@ -84,6 +107,19 @@ void BinaryFilter::initializeFilter(
...
+    change_parameters_clients_.push_back(change_parameters_client_);
+  }
```
[the remark this answers is not in the document]
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:132]: Did, I have to figure out what do if the service is not available (comment [here](https://github.com/ros-planning/navigation2/pull/3796#discussion_r1398476277))
https://github.com/ros-planning/navigation2/pull/3796/commits/496ce37c1a6506bd542017031ba2226575220527

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_filters/binary_filter.hpp (hunk 11 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_filters/binary_filter.hpp:141]: I think he was intuiting. You have a couple more linting failures https://app.circleci.com/pipelines/github/ros-planning/navigation2/10484/workflows/affb9099-bea7-40c7-b616-6c4316ee29c1/jobs/33103/tests then good to go

Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_filters/binary_filter.hpp:141]: Yes, that was some kind of intuitive suggestion for better readability. Anyway, since we need to follow linting guidelines, this could be remained w/o modification

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp]: There is no need to add global timeouts and try to re-send the service again. This should be a responsibility of DDS implementation, and if some node is not online for some reasons (e.g. bad network connection), DDS to re-try to deliver the request to the given node within a timeout. So, there is no need to do the same job twice.

Regarding throwing an exception, this is a good question. Changing the state of node - is a serious situation that might break full logic working mechanisms on the end-developer side. Imagine, if you have some handling "NodeA" that is having two sub-algorithms: robot local planning and hand manipulation. By moving into some marked area, the BinaryFilter decides to stop the robot and send a binary state to "NodeA" to switch algorithm from navigation -> to hand manipulation. If this signal won't be delivered to "NodeA", it will continue to working as previously, so the normal operation logic will break.
Additionally, this situation could be treated as ROS-parameter setting error, which is also always going with exceptions throwing.

So from this, I'd more inclined that we need to throw an error in service failure situation (and thus, stop normal BF execution).

**Review by AlexeyMerzlyakov — COMMENTED** [review]
There are a few items remained:
- Throw an exception if service was failed (discussable)
- Remove unnecessary calling of `wait_for_service` from `changeParameters()`
- Allow to use relative namespaces in ROS-parameters (https://github.com/ros-planning/navigation2/pull/3796#discussion_r1317581467)
- Switch to use non-default ROS-parameters instead of empty strings
- Future destroyed after end-of-loop question (https://github.com/ros-planning/navigation2/pull/3796#discussion_r1372142376)
- Other minor changes / nitpicks

Test code coverage is `>95%` for `binary_filter.cpp` which is OK (some exceptions were not covered):
Screenshot 2023-11-21 at 14-28-52 LCOV - bf info - plugins_costmap_filters

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp]: I mean, I agree a log is not enough, while an exception may be too much in some cases. What do you think of adding a parameter for letting the user decide this one?

Each parameter can have a parameter for this: something like a boolean `critical` (default `True`).
Any suggestion on the name is appreciated! It could be reverse also, like `is_optional` (default `False`).
 Configuration would be:
 ```yaml
     ParamNamespace:
       node_name: "node"
       param_name: "param"
       default_state: True
       is_critical: False
 ```
Hence the code should look like this:
```c++
    if (!change_parameters_client->wait_for_service(
        std::chrono::milliseconds(change_parameter_timeout_)))
    {
      if (param.is_critical){
        throw std::runtime_error("BinaryFilter: Service " + 
          std::string(change_parameters_client->get_service_name()) + 
          "not available!");
      }
      RCLCPP_ERROR(
        logger_, "BinaryFilter:  service %s not available. Skipping ...",
        change_parameters_client->get_service_name());
    } else {
      RCLCPP_INFO(
        logger_, "BinaryFilter:  service %s available.",
        change_parameters_client->get_service_name());
    }
  }
  ```

We could also think of having just one global parameter for the filter, without giving the chance to specify the behavior for each parameter, but I don't see great advantages a part from having less parameters.

What do you think?

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 9 above)
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp]: Solved with `spin_until_future_complete`

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 6 above)
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp]: Done! Thank you

**Comment by enricosutera** [comment]
> There are a few items remained:
> 
>     * Throw an exception if service was failed (discussable)
> 
>     * Remove unnecessary calling of `wait_for_service` from `changeParameters()`
> 
>     * Allow to use relative namespaces in ROS-parameters ([[binary filter] change boolean ros params #3796 (comment)](https://github.com/ros-planning/navigation2/pull/3796#discussion_r1317581467))
> 
>     * Switch to use non-default ROS-parameters instead of empty strings
> 
>     * Future destroyed after end-of-loop question ([[binary filter] change boolean ros params #3796 (comment)](https://github.com/ros-planning/navigation2/pull/3796#discussion_r1372142376))
> 
>     * Other minor changes / nitpicks from the items below
> 
> 
> Test code coverage is `>95%` for `binary_filter.cpp` which is OK (some exceptions were not covered): Screenshot 2023-11-21 at 14-28-52 LCOV - bf info - plugins_costmap_filters

@AlexeyMerzlyakov All the above points are satisfied a part from these two:

1. Just waiting for an answer on [this](https://github.com/ros-planning/navigation2/pull/3796#discussion_r1406747851)
2. Concerning this:

>     * Allow to use relative namespaces in ROS-parameters ([[binary filter] change boolean ros params #3796 (comment)](https://github.com/ros-planning/navigation2/pull/3796#discussion_r1317581467))
have you read [my comment](https://github.com/ros-planning/navigation2/pull/3796#discussion_r1398468284)? Do you have any suggestion?

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (enricosutera) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:130]: To be honest I'm not sure about how I could solve this... As I said, the filter has not the access level to distinguish between the two namespaces, so only workarounds come to my mind (e.g. the only two sub_namespace I can image for these filter are `global_costmap` and `local_costmap`, so I could simply cut them from the service name), but I don't think that  having earthbound reasoning is acceptable

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp]: What is the example when we have the parameter of some node not changed, but we are still normally operating? Is this worth to skip this node in the config file at all, which is not reliable and not necessary for us?
I doubt about such situation: imagine - you have some node which might be offline or online. And you can not guarantee it's parameters consistency. Is it real worth to support? In this case, I believe it is better to throw to developer an exception causing an error, allowing the developer to remove this unreliable node from config at all. Since the node is unreliable, it's binary param won't be guaranteed to be consistent, and thus I do not think, this is normal situation.

In other words, not-changing / not-setting of ROS parameters - is already a critical situation which throws and error in ROS2 API. For these cases, we could compare setting the parameter through a service request and thus it could be treated with exception. 
Adding additional parameter to choose throw exceptions or not for the unreliable nodes, in my opinion - over-complicating the structure, since we are still could not rely on this flaky node parameter, so why don't throw error instead in all times?

File: nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp:130]: What will be happened, if you will just remain the
```
 auto change_parameters_client_ = node->create_client<rcl_interfaces::srv::SetParameters>(
      param.node_name + "/set_parameters");
```
without forwarding "/"?

- If `param.node_name` is specified in absolute path (e.g. `/some_node`) will be will applied as `/some_node/set_parameters` service.
- If `param.node_name` is specified in relative path (e.g. `some_node`), won't service will be created relative to current BinaryFilter namespace: `some_node/set_parameters` which will be treated as `/my_namespace/some_node/set_parameters` when operating with it in run-time?

**Comment by AlexeyMerzlyakov** [comment]
@enricosutera, apologies for late respond: I was overkilled with other activities on prev. two weeks. Please refer to my comments on remaining two questions (exceptions and namespaces in service calls) from above

**Comment by SteveMacenski** [comment]
@enricosutera is it your intent to finish up this PR? I generally don't like to leave things hanging like this indefinitely

**Comment by enricosutera** [comment]
Thank you for pinging. Yes I do. I will give a last shot asap

**Comment by SteveMacenski** [comment]
Closing due to inactivity and major conflicts.
